### ISSUE TITLE
Hitch every 60 seconds when getting users from web

### ISSUE THREAD

**Issue by Rusty01** [Person8880/Shine#858]
Hi Person8880,
Your favorite server OP here :)
So within the BaseConfig.json I have it setup to use a `Refresh Interval` of 60 seconds and `get users from web`.
I have seven servers running. The problem is I have a very large permissions file.

Every 60 seconds the game lags for a split second as it gets the update. You feel it in game, which is bad.
Is there a way to have it pull from the web only on map change and server startup? and then read from that file every 60 seconds? It would only need to read the userConfig.json and not write to it, or would it still hitch?

If not, what would be the disadvantage of having it update every 15 minutes? I saw you said it should be as low as possible. 

Thank you again for all your work and dedication to keeping Shine operating! Much appreciated! <3 
~Rusty

**Comment by Person8880** [comment]
My initial guess is that the hitch may be due to saving the large file to disk (which it does to act as a fallback in case the web server is unavailable at some point). The parsing of it is staggered to avoid trying to parse everything in one go, but the saving is not. How large is the file on disk?

As for the frequency, if waiting until a map change is acceptable then you can disable the automatic refresh (set `RefreshUsers` to `false`) and it'll just load the users whenever the map changes (including server startup) which should avoid any hitching at the cost of not updating the user data as often. If you don't change the user data that often then this should be fine, refreshing is only needed to pick up changes you make, it's not needed if the data stays the same. Increasing the interval would reduce the frequency of the hitches, though not eliminate them. The wiki page was meant to imply to _not_ set the interval too low because it can be a somewhat costly operation.

**Comment by Rusty01** [comment]
Thanks, The userConfig is 2.1MB One of the reasons why it's so big is due to having a selection of 150 badges. There's no way to allow a user to select what badge they want and to decide where it goes unless you create an array as such. So the more badges we add, the bigger the config gets. I also need to optimize my UserConfig. I started doing that last night to take advantage of the inheritance options within the userConfig which I wasn't doing before. I did just set the `RefreshUsers` to `false` I don't change the user data that often so that information was a big help, thanks. 

One last question, in my userconfig I have a group, within the group I have badges. The badges are listed as such.
slot1: redbadge, bluebadge, yellowbadge
slot2: redbadge, bluebadge, yellowbadge
etc.
Is it possible to only list the name of the badge once but have it available to all badge slots? Or is there no way around this?
Thanks again for all your help! 


slot1: redbadge, bluebadge, yellowbadge
slot2: redbadge, bluebadge, yellowbadge

**Comment by Person8880** [comment]
At the moment there's no way to avoid the slot duplication. The format could be improved to try and reduce the duplication, though it feels like the badge system itself should just drop the notion of individual slot badge lists as it's just awkward, but that's the game's responsibility.

For the inheritance, badges are added together from inherited groups, so in your example the admin group would have access to both the red and green badges. However, you can [specify `ForcedBadges` on the group](https://github.com/Person8880/Shine/wiki/Badges#forcing-badges) to force slot 1 to be the badge you want for them, which means even though multiple badges show up in the menu, they can't use the others.

**Comment by Person8880** [comment]
I've made some improvements to the badge handling to reduce its performance impact further with lots of badges. Turns out that each column for each badge triggers a network message for a connected player, and if you have 150 x 8 badges, that's 1200 network messages sent to a player which could have been why there's a significant hitch. I've made the badges get pre-configured at startup before any players join which eliminates these network messages as long as the user data isn't reloaded.

I've also added the ability to created named badge lists in the user config using the existing master badge table. See [the wiki for details](https://github.com/Person8880/Shine/wiki/Badges#named-badge-lists).

These changes will be in the next update (which is unlikely to be this week, possibly next week).
